### PR TITLE
refactor: Utility for grabbing all configs

### LINKED ISSUE #11852: Schematics util to fetch all spartacus configs from client app
Why?
In multiple places we need to check if we already have that config, or should we add it.

Eg. we should not duplicate configs for registering translations.

### PR DESCRIPTION
**Pull request by marlass**
Closes #11852

### CODE REVIEW

File: projects/schematics/src/shared/utils/config-utils.ts [hunk 1]
```diff
@@ -296,3 +300,60 @@ function convert(newValues: string | string[]): string {
...
+  const configs: Node[] = [];
+  const module = getModule(sourceFile);
+  if (!module) {
+    return [];
+  }
+  const literal = module.getFirstDescendantByKind(
+    tsMorph.SyntaxKind.ObjectLiteralExpression
+  );
+  if (literal) {
+    const properties = literal.getChildrenOfKind(
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils.ts]: Let's try to minimize the amount of indentations :)
We could do `const properties = literal.getChildrenOfKind() ?? []` and avoid the `if (properties) {`.

File: projects/schematics/src/shared/utils/config-utils.ts [hunk 2]
```diff
@@ -296,3 +300,60 @@ function convert(newValues: string | string[]): string {
   }
   return configValue;
 }
+
+export function getConfigs(sourceFile: SourceFile): Node[] {
+  const configs: Node[] = [];
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils.ts]: We can declare this variable just before the line 313

File: projects/schematics/src/shared/utils/config-utils.ts [hunk 3]
```diff
@@ -296,3 +300,60 @@ function convert(newValues: string | string[]): string {
...
+          property.getInitializerIfKind(
+            tsMorph.SyntaxKind.ArrayLiteralExpression
+          )
+        ) {
+          const initializer = property.getInitializerIfKind(
+            tsMorph.SyntaxKind.ArrayLiteralExpression
+          );
+          if (initializer) {
+            const elements = initializer.getElements();
+            elements.forEach((element) => {
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils.ts]: We can also streamline this to `initializer?.getElements().forEach((element) => {` and avoid having the nested `if` statement.

File: projects/schematics/src/shared/utils/new-module-utils.ts [hunk 4]
```diff
@@ -212,3 +212,25 @@ function elementExists(
 
   return false;
 }
+
+export function getModule(sourceFile: SourceFile): CallExpression | undefined {
```
Comment (znikola) [inline projects/schematics/src/shared/utils/new-module-utils.ts:207]: Can we refactor the `addToModuleInternal()` from this file to use this new helper function?

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/schematics/src/shared/utils/import-utils.ts [hunk 5]
```diff
@@ -18,3 +18,21 @@ export function isImportedFrom(node: Identifier, importPath: string): boolean {
...
+export function isImportedFromSpartacusLibs(node: Identifier): boolean {
+  let result = false;
+  const definitions = node.getDefinitions();
+  for (const def of definitions) {
+    const node = def.getDeclarationNode();
+
+    const declaration = node?.getFirstAncestorByKind(
+      ts.SyntaxKind.ImportDeclaration
+    );
+    if (declaration?.getModuleSpecifier().getText().includes('@spartacus/')) {
```
Comment (znikola) [inline projects/schematics/src/shared/utils/import-utils.ts]: Maybe it's safer to use `startsWith`.

File: projects/schematics/src/shared/utils/import-utils.ts [hunk 6]
```diff
@@ -18,3 +18,21 @@ export function isImportedFrom(node: Identifier, importPath: string): boolean {
...
+      result = true;
```
Comment (znikola) [inline projects/schematics/src/shared/utils/import-utils.ts]: You could just `return true` here, get rid of the `let result = false;` and `return false` at the end of the method.
It's a personal preference, and your call if you want to do it.

File: projects/schematics/src/shared/utils/config-utils_spec.ts [hunk 7]
```diff
@@ -200,61 +201,82 @@ describe('Storefront config utils', () => {
...
+provideConfig(<TmsConfig>{
+  tagManager: {
+    gtm: {
+      events: [NavigationEvent, CartAddEntrySuccessEvent],
+    },
+    aep: {
+      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
+    },
+  },
+}),
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils_spec.ts:238]: Does it make sense to repeat this configuration, but using the `as` syntax? In this way, we are confident that we covering both cases.

File: projects/schematics/src/shared/utils/config-utils.ts [hunk 8]
```diff
@@ -349,3 +349,40 @@ export function getConfigs(sourceFile: SourceFile): Node[] {
   }
   return configs;
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils.ts]: After thinking a bit, let's remove this method. No need to keep it if it's not used anywhere.

File: projects/schematics/src/shared/utils/config-utils_spec.ts [hunk 9]
```diff
@@ -200,61 +201,82 @@ describe('Storefront config utils', () => {
...
-  ],
+imports: [
+BaseTmsModule.forRoot(),
+GtmModule,
+AepModule,
+PersonalizationRootModule,
+],
+providers: [
+someFunction(),
+...someSpread,
```
Comment (znikola) [inline projects/schematics/src/shared/utils/config-utils_spec.ts:227]: Nice.

**Review by znikola — CHANGES_REQUESTED** [review]

File: projects/schematics/src/shared/utils/config-utils_spec.ts (hunk 7 above)
Comment (marlass) [reply projects/schematics/src/shared/utils/config-utils_spec.ts:238]: Would make sense if we would keep the config util ;)

**Review by znikola — APPROVED** [review]

**Pull request merged**